### PR TITLE
Add OTel collector on Agent Startup based on configuration

### PR DESCRIPTION
**Pull request by oliveromahony**
### Proposed changes

Add an embedded OTel Collector to NGINX Agent

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [`CONTRIBUTING`](https://github.com/nginx/agent/blob/main/docs/CONTRIBUTING.md) document
- [x] I have run ```make install-tools``` and have attached any dependency changes to this pull request
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works
- [x] If applicable, I have checked that any relevant tests pass after adding my changes
- [ ] If applicable, I have updated any relevant documentation ([`README.md`](https://github.com/nginx/agent/blob/main/README.md))
- [ ] If applicable, I have tested my cross-platform changes on Ubuntu 22, Redhat 8, SUSE 15 and FreeBSD 13

### CODE REVIEW

File: internal/collector/config_test.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	assert.NotNil(t, settings.ResolverSettings, "ResolverSettings should not be nil")
+	assert.Len(t, settings.ResolverSettings.ProviderFactories, 5, "There should be 5 provider factories")
+	assert.Len(t, settings.ResolverSettings.ConverterFactories, 1, "There should be 1 converter factory")
+	assert.NotEmpty(t, settings.ResolverSettings.URIs, "URIs should not be empty")
+	assert.Equal(t, "/tmp/otel-collector-config.yaml", settings.ResolverSettings.URIs[0], "Default URI should match")
+}
+
+func TestGetConfig(t *testing.T) {
+	// Test with environment variable set
+	os.Setenv("OPENTELEMETRY_COLLECTOR_CONFIG_FILE", "/path/to/config.yaml")
```
Comment (olli-holmala) [inline internal/collector/config_test.go:41]: I think `testing.T` has a `Setenv` function
```suggestion
	t.Setenv("OPENTELEMETRY_COLLECTOR_CONFIG_FILE", "/path/to/config.yaml")
```

File: internal/collector/factories.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+	"go.opentelemetry.io/collector/exporter/otlphttpexporter"
+	"go.opentelemetry.io/collector/extension"
+	"go.opentelemetry.io/collector/otelcol"
+	"go.opentelemetry.io/collector/processor"
+	"go.opentelemetry.io/collector/processor/batchprocessor"
+	"go.opentelemetry.io/collector/processor/memorylimiterprocessor"
+	"go.opentelemetry.io/collector/receiver"
+	"go.opentelemetry.io/collector/receiver/nopreceiver"
+	"go.opentelemetry.io/collector/receiver/otlpreceiver"
+	"go.uber.org/multierr"
```
Comment (olli-holmala) [inline internal/collector/factories.go:79]: I think this is no longer needed, as Go now has `errors.Join()` which does the same thing as `multierr`

File: internal/plugin/otel_collector.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	// Collector The OTel collector plugin start an embedded OTel collector for metrics collection in the OTel format.
+	Collector struct {
+		svc     *otelcol.Collector
+		appDone chan struct{}
+		stopped bool
+		cancel  context.CancelFunc
+		config  *config.Config
+	}
+)
+
```
Comment (olli-holmala) [inline internal/plugin/otel_collector.go:28]: Should verify `bus.Plugin` compliance here.

File: internal/plugin/otel_collector.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+func (oc *Collector) Init(ctx context.Context, mp bus.MessagePipeInterface) error {
+	slog.InfoContext(ctx, "Starting OTel Collector plugin")
+
+	var runCtx context.Context
+	runCtx, oc.cancel = context.WithCancel(ctx)
+
+	go func() {
+		err := oc.run(runCtx)
+		if err != nil {
+			slog.ErrorContext(ctx, "error", err)
```
Comment (olli-holmala) [inline internal/plugin/otel_collector.go:53]: Should this be `runCtx`?
```suggestion
			slog.ErrorContext(runCtx, "error", err)
```

File: internal/plugin/otel_collector.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+		if appErr != nil {
+			err = appErr
+			slog.ErrorContext(ctx, "error", appErr)
+		}
+	}()
+
+	for {
+		state := oc.svc.GetState()
+
+		oc.svc.GetState()
```
Comment (olli-holmala) [inline internal/plugin/otel_collector.go:76]: Why is `GetState()` called twice?

File: internal/plugin/otel_collector.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+		case otelcol.StateRunning:
+			return nil
+		case otelcol.StateClosing, otelcol.StateClosed:
+		default:
+			err = fmt.Errorf("unable to start, otelcol state is %d", state)
+		}
+	}
+}
+
+// Info about the plugin. Required for the `Plugin` interface.
```
Comment (olli-holmala) [inline internal/plugin/otel_collector.go:95]: We can remove these 'Required for the `Plugin` interface.' comments after interface compliance is verified.

File: internal/plugin/plugin_manager.go [hunk 7]
```diff
@@ -78,3 +79,16 @@ func isGrpcClientConfigured(agentConfig *config.Config) bool {
 		agentConfig.Command.Server != nil &&
 		agentConfig.Command.Server.Type == "grpc"
 }
+
+func addCollector(agentConfig *config.Config, logger *slog.Logger, plugins []bus.Plugin) []bus.Plugin {
+	if agentConfig.Metrics.Collector {
```
Comment (olli-holmala) [inline internal/plugin/plugin_manager.go:84]: Missing a nil check for `agentConfig.Metrics`

**Review by olli-holmala — APPROVED** [review]
LGTM 👍

File: internal/plugin/otel_collector.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	"github.com/nginx/agent/v3/internal/bus"
+	"github.com/nginx/agent/v3/internal/collector"
+	"github.com/nginx/agent/v3/internal/config"
+	"go.opentelemetry.io/collector/otelcol"
+)
+
+type (
+	// Collector The OTel collector plugin start an embedded OTel collector for metrics collection in the OTel format.
+	Collector struct {
+		svc     *otelcol.Collector
```
Comment (dhurley) [inline internal/plugin/otel_collector.go:21]: whats `svc`?

File: internal/plugin/otel_collector.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+func (oc *Collector) run(ctx context.Context) error {
+	var err error
+	oc.appDone = make(chan struct{})
+
+	go func() {
+		defer close(oc.appDone)
+		appErr := oc.svc.Run(ctx)
+		if appErr != nil {
+			err = appErr
+			slog.ErrorContext(ctx, "error", appErr)
```
Comment (dhurley) [inline internal/plugin/otel_collector.go:71]: No need to log error if its being returned

**Review by dhurley — COMMENTED** [review]
Should everything here be in a `metrics` package? We want to move away from having a `plugins` package so can we move `plugin/otel_collector.go` to the `collector` package and rename it to `otel_collector_plugin.go`?

File: internal/plugin/otel_collector.go (hunk 8 above)
Comment (oliveromahony) [reply internal/plugin/otel_collector.go:21]: the running service? Collector.collector seemed incorrect. What would you prefer?

Comment (dhurley) [reply internal/plugin/otel_collector.go]: oh right it wasn't obvious to me what `svc` meant. Could it be called just `service`?

**Review by aphralG — APPROVED** [review]

**Pull request merged**